File: Frontend/services/api_client.py
import os
import requests
import streamlit as st
from typing import Any, Dict

BASE_URL = "http://localhost:8000/api"

def get_headers():
    token = st.session_state.get('id_token')
    if not token:
        raise ValueError("Not authenticated")
    return {"Authorization": f"Bearer {token}"}

def generate_blueprint(brief: Dict[str, Any]) -> Dict[str, Any]:
    url = f"{BASE_URL}/generate"
    cloud_val = str(brief.get("cloud", "aws")).lower()
    if cloud_val not in ["aws", "azure", "gcp", "none"]:
        cloud_val = "aws"
    try:
        traffic_val = int(brief.get("traffic", 1000))
    except (ValueError, TypeError):
        traffic_val = 1000
    try:
        timeline_val = int(brief.get("timeline", 3))
    except (ValueError, TypeError):
        timeline_val = 3

    payload = {
        "business_idea": brief.get("idea", "").strip(),
        "tech_preference": "enterprise" if brief.get("technology", "").lower() == "enterprise" else "opensource",
        "cloud_preference": cloud_val,
        "expected_daily_traffic": traffic_val,
        "delivery_timeline_months": timeline_val,
        "country": str(brief.get("country", "US")).strip() or "US"
    }
    resp = requests.post(url, json=payload, headers=get_headers())
    resp.raise_for_status()
    data = resp.json()
    blueprint_data = data.get("blueprint", {})
    # Mangle response back to frontend expected structure
    return {
        'summary': blueprint_data.get("requirements", {}).get("summary", ""),
        'scope': blueprint_data.get("requirements", {}).get("scope", []),
        'brief': brief,
        'stack': [
            ["Frontend", blueprint_data.get("technology", {}).get("frontend", "")],
            ["Backend", blueprint_data.get("technology", {}).get("backend", "")],
            ["Database", blueprint_data.get("technology", {}).get("database", "")],
            ["Cloud", blueprint_data.get("technology", {}).get("cloud", "")]
        ],
        'risks': [
            [risk.get("risk", ""), risk.get("impact", ""), risk.get("mitigation", "")]
            for risk in blueprint_data.get("validation", {}).get("risks", []) if isinstance(risk, dict)
        ],
        'project_id': blueprint_data.get("project_id"),
        'generation_id': blueprint_data.get("generation_id")
    }

def list_projects() -> list:
    url = f"{BASE_URL}/projects"
    resp = requests.get(url, headers=get_headers())
    resp.raise_for_status()
    return resp.json()

def get_project(project_id: str) -> dict:
    url = f"{BASE_URL}/projects/{project_id}"
    resp = requests.get(url, headers=get_headers())
    resp.raise_for_status()
    return resp.json()

File: Frontend/src/services/apiClient.ts
import { getAuth } from 'firebase/auth';

const BASE_URL = (import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000/api').replace(/\/$/, '');

async function authenticatedFetch(path: string, init: RequestInit = {}) {
  const user = getAuth().currentUser;
  if (!user) throw new Error('Please sign in again.');
  const token = await user.getIdToken();
  const headers = new Headers(init.headers);
  headers.set('Authorization', `Bearer ${token}`);
  const response = await fetch(`${BASE_URL}${path}`, { ...init, headers });
  if (!response.ok) {
    const body = await response.json().catch(() => null);
    throw new Error(body?.error?.message || body?.detail || `Request failed (${response.status})`);
  }
  return response;
}

export type GenerateRequest = {
  business_idea: string; tech_preference: 'opensource' | 'enterprise';
  cloud_preference: 'aws' | 'azure' | 'gcp' | 'none'; expected_daily_traffic: number;
  delivery_timeline_months: number; country: string;
};

export async function generateBlueprintAPI(requestData: GenerateRequest) {
  return (await authenticatedFetch('/generate', { method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(requestData) })).json();
}
export async function getProjectsAPI() { return (await authenticatedFetch('/projects')).json(); }
export async function getProjectDetailAPI(projectId: string) { return (await authenticatedFetch(`/projects/${projectId}`)).json(); }
export async function getGenerationDetailAPI(generationId: string) { return (await authenticatedFetch(`/generations/${generationId}`)).json(); }
export async function getUserProfileAPI() { return (await authenticatedFetch('/users/me')).json(); }
export async function updateUserProfileAPI(name: string, photo_url = '') {
  return (await authenticatedFetch('/users/me', { method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify({ name, photo_url }) })).json();
}
export async function downloadReport(generationId: string, kind: 'report' | 'pdf') {
  const response = await authenticatedFetch(`/generations/${generationId}/${kind}`);
  const blob = await response.blob();
  const url = URL.createObjectURL(blob);
  const link = document.createElement('a');
  link.href = url; link.download = `solutionforge-${generationId}.${kind === 'pdf' ? 'pdf' : 'html'}`;
  document.body.appendChild(link); link.click(); link.remove(); URL.revokeObjectURL(url);
}
